refactor: generate openapi directly instead of legacy format

d11n opened the pull request:
fixes #481 

removing the legacy format entirely means adding/changing features will be done directly to the openapi document, instead of first modifying the legacy route format and then converting that to an openapi operation...a handful of issues will be quick to resolve once this is merged...the goal of this pr is to get to 100% tests passing without changing any openapi docs, just the doc generation algorithm

d11n commented:
at present:
```bash
 PASS  test/unit/options-urls-test.js 9 OK 818ms
 PASS  test/integration/landing-page-test.js 1 OK 1s
 FAIL  test/integration/endpoints-test.js 21s
Suites:   1 failed, 2 passed, 3 of 3 completed
Asserts:  378 failed, 634 passed, of 1012
Time:     21s
```
current failures are primarily around operation sorting and `x-changes`

gr2m commented on lib/endpoint/find-accept-header.js:
```diff
@@ -0,0 +1,28 @@
+  const previewBlocks = state.blocks.filter(block => block.type === 'preview')
+
+  state.routes.forEach(route => {
+    // TODO: handle optional preview blocks
```
I think this comment is obsolete and can be removed

d11n replied:
currently, there are 73 routes on `api.github.com` that come through with `previewBlocks` that have `required: false`

- https://developer.github.com/v3/apps/installations/#list-repositories
- https://developer.github.com/v3/issues/#list-issues
- https://developer.github.com/v3/orgs/#get-an-organization
- https://developer.github.com/v3/orgs/#edit-an-organization
- https://developer.github.com/v3/orgs/members/#list-pending-organization-invitations
- https://developer.github.com/v3/orgs/members/#create-organization-invitation
- https://developer.github.com/v3/orgs/members/#list-organization-invitation-teams
- https://developer.github.com/v3/repos/#list-organization-repositories
- https://developer.github.com/v3/repos/#create
- https://developer.github.com/v3/teams/#list-teams
- https://developer.github.com/v3/teams/#create-team
- https://developer.github.com/v3/repos/#edit
- https://developer.github.com/v3/repos/branches/#list-team-restrictions-of-protected-branch
- https://developer.github.com/v3/repos/branches/#replace-team-restrictions-of-protected-branch
- https://developer.github.com/v3/repos/branches/#add-team-restrictions-of-protected-branch
- https://developer.github.com/v3/repos/branches/#remove-team-restrictions-of-protected-branch
- https://developer.github.com/v3/repos/collaborators/#list-collaborators
- https://developer.github.com/v3/repos/collaborators/#check-if-a-user-is-a-collaborator
- https://developer.github.com/v3/repos/comments/#list-commit-comments-for-a-repository
- https://developer.github.com/v3/repos/comments/#get-a-single-commit-comment
- https://developer.github.com/v3/repos/comments/#list-comments-for-a-single-commit
- https://developer.github.com/v3/repos/deployments/#list-deployments
- https://developer.github.com/v3/repos/deployments/#create-a-deployment
- https://developer.github.com/v3/repos/deployments/#get-a-single-deployment
- https://developer.github.com/v3/repos/deployments/#list-deployment-statuses
- https://developer.github.com/v3/repos/deployments/#create-a-deployment-status
- https://developer.github.com/v3/repos/deployments/#get-a-single-deployment-status
- https://developer.github.com/v3/issues/#list-issues-for-a-repository
- https://developer.github.com/v3/issues/#create-an-issue
- https://developer.github.com/v3/issues/comments/#list-comments-in-a-repository
- https://developer.github.com/v3/issues/comments/#get-a-single-comment
- https://developer.github.com/v3/issues/events/#list-events-for-a-repository
- https://developer.github.com/v3/issues/events/#get-a-single-event
- https://developer.github.com/v3/issues/#get-a-single-issue
- https://developer.github.com/v3/issues/#edit-an-issue
- https://developer.github.com/v3/issues/assignees/#add-assignees-to-an-issue
- https://developer.github.com/v3/issues/assignees/#remove-assignees-from-an-issue
- https://developer.github.com/v3/issues/comments/#list-comments-on-an-issue
- https://developer.github.com/v3/issues/events/#list-events-for-an-issue
- https://developer.github.com/v3/issues/labels/#list-labels-on-an-issue
- https://developer.github.com/v3/issues/labels/#add-labels-to-an-issue
- https://developer.github.com/v3/issues/labels/#replace-all-labels-for-an-issue
- https://developer.github.com/v3/issues/labels/#remove-a-label-from-an-issue
- https://developer.github.com/v3/issues/#lock-an-issue
- https://developer.github.com/v3/issues/timeline/#list-events-for-an-issue
- https://developer.github.com/v3/issues/labels/#list-all-labels-for-this-repository
- https://developer.github.com/v3/issues/labels/#create-a-label
- https://developer.github.com/v3/issues/labels/#get-a-single-label
- https://developer.github.com/v3/issues/labels/#get-labels-for-every-issue-in-a-milestone
- https://developer.github.com/v3/repos/pages/#get-information-about-a-pages-site
- https://developer.github.com/v3/repos/pages/#enable-a-pages-site
- https://developer.github.com/v3/repos/pages/#update-information-about-a-pages-site
- https://developer.github.com/v3/pulls/#list-pull-requests
- https://developer.github.com/v3/pulls/#create-a-pull-request
- https://developer.github.com/v3/pulls/comments/#list-comments-in-a-repository
- https://developer.github.com/v3/pulls/comments/#get-a-single-comment
- https://developer.github.com/v3/pulls/#get-a-single-pull-request
- https://developer.github.com/v3/pulls/#update-a-pull-request
- https://developer.github.com/v3/pulls/comments/#list-comments-on-a-pull-request
- https://developer.github.com/v3/pulls/review_requests/#create-a-review-request
- https://developer.github.com/v3/repos/#list-all-topics-for-a-repository
- https://developer.github.com/v3/repos/#replace-all-topics-for-a-repository
- https://developer.github.com/v3/search/#search-issues-and-pull-requests
- https://developer.github.com/v3/search/#search-labels
- https://developer.github.com/v3/search/#search-repositories
- https://developer.github.com/v3/search/#search-topics
- https://developer.github.com/v3/teams/#get-team
- https://developer.github.com/v3/teams/#edit-team
- https://developer.github.com/v3/teams/#delete-team
- https://developer.github.com/v3/teams/discussions/#list-discussions
- https://developer.github.com/v3/teams/discussions/#create-a-discussion
- https://developer.github.com/v3/teams/discussions/#get-a-single-discussion
- https://developer.github.com/v3/teams/discussions/

d11n replied:
sorry for that long comment...the non-required accept headers for those routes are being ignored right now...after a cursory look, they seem to generally be used for adding extra info in the response...which if that's true, we could just always just get everything...but some analysis may be necessary to figure out exactly how to treat these, and even ensure if they're being parsed correctly as `required: false`

d11n commented:
^
```
222 failed, 790 passed, of 1012
```

d11n commented:
14 failures left that are not related to `x-changes` or `x-code-samples`

d11n commented:
36 failures in `x-code-samples` and `x-changes` is still always empty

d11n commented:
only thing left is adding `x-changes` (114 failures)

d11n commented:
present state of things:

---

**`$ git status`**
```
On branch d11n/481-remove-legacy
Your branch is up to date with 'origin/d11n/481-remove-legacy'.

nothing to commit, working tree clean
```

**`$ npm run routes:lint`**
```
> @octokit/routes@0.0.0-development routes:lint /home/derek/code/octokit/routes
> speccy lint openapi/api.github.com/index.json --skip openapi-tags

Specification is valid, with 0 lint errors
```

**`$ npm run test:ci`**
```
> @octokit/routes@0.0.0-development test:ci /home/derek/code/octokit/routes
> node bin/run-tests.js

TAP version 13
...
ok 1 - unit tests # time=999.07ms
...
ok 2 - api.github.com integration tests # time=10187.444ms
...
ok 3 - ghe-2.15 integration tests # time=9856.876ms
...
ok 4 - ghe-2.16 integration tests # time=9742.17ms
...
ok 5 - ghe-2.17 integration tests # time=10246.14ms

1..5
# time=41091.007ms
```

---

**`$ git status`**
```
On branch d11n/481-remove-legacy
Your branch is up to date with 'origin/d11n/481-remove-legacy'.

nothing to commit, working tree clean
```

**`$ npm run update -- --cached`**
```
> @octokit/routes@0.0.0-development update /home/derek/code/octokit/routes
> node bin/octokit-rest-routes.js update "--cached"

🤖  Looking for sections at 755 URLs
...
🏁  done
```

**`$ git status`**
```
On branch d11n/481-remove-legacy
Your branch is up to date with 'origin/d11n/481-remove-legacy'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   openapi/api.github.com/operations/teams/add-member.json
	modified:   openapi/api.github.com/operations/teams/get-member.json
	modified:   openapi/api.github.com/operations/teams/remove-member.json

no changes added to commit (use "git add" and/or "git commit -a")
```

**`$ git diff`** `openapi/api.github.com/operations/teams/` **`add-member.json`**
```
diff --git a/openapi/api.github.com/operations/teams/add-member.json b/openapi/api.github.com/operations/teams/add-member.json
index e5efabd5f..4323d056f 100644
--- a/openapi/api.github.com/operations/teams/add-member.json
+++ b/openapi/api.github.com/operations/teams/add-member.json
@@ -58,5 +58,6 @@
     "enabledForApps": true,
     "githubCloudOnly": false
   },
-  "x-changes": []
+  "x-changes": [],
+  "deprecated": true
 }
```

**`$ git diff`** `openapi/api.github.com/operations/teams/` **`get-member.json`**
```
diff --git a/openapi/api.github.com/operations/teams/get-member.json b/openapi/api.github.com/operations/teams/get-member.json
index b2dd5dbb9..e399859a4 100644
--- a/openapi/api.github.com/operations/teams/get-member.json
+++ b/openapi/api.github.com/operations/teams/get-member.json
@@ -58,5 +58,6 @@
     "enabledForApps": true,
     "githubCloudOnly": false
   },
-  "x-changes": []
+  "x-changes": [],
+  "deprecated": true
 }
```

**`$ git diff`** `openapi/api.github.com/operations/teams/` **`remove-member.json`**
```
diff --git a/openapi/api.github.com/operations/teams/remove-member.json b/openapi/api.github.com/operations/teams/remove-member.json
index 714319fbd..2ea7d432f 100644
--- a/openapi/api.github.com/operations/teams/remove-member.json
+++ b/openapi/api.github.com/operations/teams/remove-member.json
@@ -58,5 +58,6 @@
     "enabledForApps": true,
     "githubCloudOnly": false
   },
-  "x-changes": []
+  "x-changes": [],
+  "deprecated": true
 }
```

---

looks like the test checks weren't triggered...at least there's no indication whether they passed or failed on the checks tab...not sure what to do about this...but it's ready for review @gr2m 

and just fyi...the gratuitous amounts of todos need not become issues...the important ones will be done in short order after this is merged...it's just helpful to know where in the code certain changes need to be made...some of them are effectively duplicates of existing issues

gr2m commented on lib/endpoint/find-parameters.js:
```diff
@@ -342,21 +359,114 @@ function mapToInput (state) {
-    required: rawParam.required,
-    description: rawParam.description || `${rawParam.name} parameter`
+function formatParam (param) {
+  const { method, path, operation } = this
```
I try to avoid `this` to avoid hard to find bugs and make the code easier to comprehend. I'd instead pre-bind the format param function

```diff
- .map(formatParam, route)
+ .map(formatParam.bind(null, route))
```

Or pass it explicitly

```diff
- .map(formatParam, route)
+ .map(rawParam => formatParam(route, rawParam))
```

Probably the latter is the most readable?

gr2m replied:
nevermind looks like you removed that in commits after :)
